File: Algorithms/1438.绝对差不超过限制的最长连续子数组/longestSubarray.java
class Solution {
    public int longestSubarray(int[] nums, int limit) {
        int left = 0, ans = 0;
        TreeMap<Integer, Integer> m = new TreeMap<>();
        for (int right = 0; right < nums.length; ++right) {
            m.put(nums[right], m.getOrDefault(nums[right], 0) + 1);
            while (m.lastKey() - m.firstKey() > limit) {
                m.put(nums[left], m.get(nums[left]) - 1);
                if (m.get(nums[left]) == 0)
                    m.remove(nums[left]);
                ++left;
            }
            ans = Math.max(ans, right - left + 1);
        }
        return ans;
    }
}

File: Algorithms/1438/longestSubarray.py
from collections import deque

class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        max_q, min_q = deque(), deque()
        left = ans = 0
        for right in range(len(nums)):
            while max_q and max_q[-1] < nums[right]:
                max_q.pop()
            max_q.append(nums[right])
            while min_q and min_q[-1] > nums[right]:
                min_q.pop()
            min_q.append(nums[right])

            while max_q[0] - min_q[0] > limit:
                if max_q[0] == nums[left]:
                    max_q.popleft()
                if min_q[0] == nums[left]:
                    min_q.popleft()
                left += 1
            ans = max(ans, right - left + 1)
        return ans
